[COREV3] Scaffold out initial shape of `@typespec/ts-http-runtime`

xirzec opened the pull request:
This PR creates the initial draft of `@typespec/ts-http-runtime` -- a new package for the "unbranded core" effort when generating SDKs from TypeSpec that are not targeting Azure services.

Morally this can be viewed as `core-rest-pipeline` and `core-client-rest` plus everything they depend on (abort-controller, core-auth, core-tracing, core-util, and logger)

I've done very little pruning of Azure-specific things outside of cleaning up the Logger names and removing some Azure-only policies.

Update: tests have also been migrated

deyaaeldeen commented on sdk/core/ts-http-runtime/tsconfig.json:
```diff
@@ -0,0 +1,12 @@
+    "outDir": "./dist-esm",
+    "declarationDir": "./types",
+    "paths": {
+      "@azure/ts-http-runtime": ["./src/index"]
```
```suggestion
      "@typespec/ts-http-runtime": ["./src/index"]
```

jeremymeng commented on sdk/core/ts-http-runtime/src/util/base64.browser.ts:
```diff
@@ -0,0 +1,45 @@
+// Copyright (c) Microsoft Corporation.
```
This is not needed? I don't see the corresponding node version

xirzec replied:
it gets imported by sha256.browser.js -- IIRC Timo has cleaned some of this up in a recent PR to core-util, we can port his changes after that PR goes in.

deyaaeldeen reviewed (commented):
I left some comments. Should we also provide an `exports` section in package.json that groups relevant bits together?

jeremymeng commented on sdk/core/ts-http-runtime/src/util/utf8.browser.ts:
```diff
@@ -0,0 +1,41 @@
+// Copyright (c) Microsoft Corporation.
```
not needed? no seeing utf8.ts

jeremymeng commented on sdk/core/ts-http-runtime/src/log.ts:
```diff
@@ -0,0 +1,5 @@
+// Copyright (c) Microsoft Corporation.
```
nti: this may be exported from logger now?

lirenhe commented on sdk/core/ts-http-runtime/package.json:
```diff
@@ -0,0 +1,136 @@
+{
+  "name": "@typespec/ts-http-runtime",
```
Can you share the thinking about using '@typespec'.  Does it mean this lib could only be used for typespec scenario?
I am wondering whether we could have a more generic name.

joheredi replied:
The primary reason for this naming convention is that this library is predominantly meant for code generated with TypeSpec. Given that TypeSpec is the language in use, adding a "runtime" suffix helps clarify that this is the actual runtime implementation. This also means we can leverage the established "TypeSpec" brand instead of creating a new one.

However, there's nothing preventing its use outside of TypeSpec code generation. While it might not be optimal in those contexts, our aim is to optimize for the majority of use cases.

The naming isn't set in stone. We're inclined towards this name for now, but we're open to suggestions. For the moment, we'll proceed with this as a starting point and make adjustments as needed in the future.

xirzec replied:
Yeah, part of the issue is we want a namespace we have control over (to prevent registry name confusion attacks) and we want to avoid 'Azure' in the branding. Potentially this could go in the `@microsoft` namespace as well?

joheredi reviewed (approved):
Thanks for putting this together. It's an excellent starting point for our generic library project.

The comments in this PR are great and I have filed issues for them which I'll track to make sure they get addressed.

As we progress, I anticipate we'll take several targeted passes to refine and optimize. For instance, we'll look to:
- Identify opportunities to converge in common APIs that are now possible in modern Node runtimes
- Identify opportunities to remove polyfills no longer needed since IE11 is no longer in the picture
- Identify opportunities to drop dependencies in favor of native APIs
- Debranding in docs, comments, and code.


I've set up a project to oversee and track our progress related to this effort. I'm still filling work items and organizing the project but you can take a peak here: https://github.com/orgs/Azure/projects/567/views/1

Getting these base changes merged is key. They'll lay the groundwork for the next steps and other work streams building on this. So, let's get this in and tweak as needed!

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[typespec-ts-http-runtime](https://apiview.dev/Assemblies/Review/8e850596e78f4b2798b9869c91e4fa86)

xirzec commented on sdk/core/ts-http-runtime/core-rest-pipeline.shims.d.ts:
```diff
@@ -0,0 +1,12 @@
+  interface Blob {}
+  interface File {}
+  interface ReadableStream<R = any> {}
+  interface TransformStream<I = any, O = any>  {}
```
this is a good point, can perhaps also remove from `core-rest-pipeline`

xirzec commented on sdk/core/ts-http-runtime/karma.conf.js:
```diff
@@ -0,0 +1,119 @@
+    // inject following environment values into browser testing with window.__env__
+    // environment values MUST be exported or set with same console running "karma start"
+    // https://www.npmjs.com/package/karma-env-preprocessor
+    envPreprocessor: ["ACCOUNT_NAME", "ACCOUNT_SAS"],
```
I don't believe so, I think this was old cruft from some of the core-http samples

xirzec commented on sdk/core/ts-http-runtime/package.json:
```diff
@@ -0,0 +1,136 @@
+    "LICENSE",
+    "README.md"
+  ],
+  "repository": "github:Azure/azure-sdk-for-js",
```
For now, I think yes, if nothing else it will greatly ease our engineering efficiency

deyaaeldeen replied:
I agree, at least until it gets used by someone external to Microsoft. At that point in time, I think we'll need to consider another home for it, to be more aligned with the "unbranded" vision.

xirzec commented:
I addressed some of the above feedback that was quick, but in the interest of forward progress, I'm going to enable auto-merge and we can continue to iterate before shipping the first beta
